src/helpers/buffers/unordered_receiver.rs

```diff
@@ -0,0 +1,502 @@
+    /// don't guarantee that all contexts are awoken.
+    ///
+    /// # Panics
+    ///
+    /// If `i` is for an message that has already been read.
+    ///
+    /// [`recv`]: UnorderedReceiver::recv
+    /// [`poll`]: Future::poll
+    fn add_waker(&mut self, i: usize, waker: Waker) {
+        // We don't save a waker at `self.next`, so `>` and not `>=`.
+        if i > self.next + self.wakers.len() {
+            self.overflow_wakers.push(waker);
+        } else {
+            assert!(
+                i > self.next,
+                "Awaiting a read that has already been fulfilled"
+            );
+            let index = i % self.wakers.len();
+            self.wakers[index] = Some(waker);
+        }
+    }
+
+    /// Wake the waker from the next future, if the next receiver has been polled.
+    fn wake_next(&mut self) {
+        self.next += 1;
+        let index = self.next % self.wakers.len();
+        if let Some(w) = self.wakers[index].take() {
+            w.wake();
+        }
+        // See the documentation for `overflow_wakers`.
+        if self.next % (self.wakers.len() / 2) == 0 {
+            for w in self.overflow_wakers.drain(..) {
+                w.wake();
+            }
+        }
+    }
+
+    /// Poll for the next record.  This should only be invoked when
+    /// the future for the next message is polled.
+    fn poll_next<M: Message>(&mut self, cx: &mut Context<'_>) -> Poll<Result<M, Error>> {
+        if let Some(m) = self.spare.read() {
+            self.wake_next();
+            return Poll::Ready(Ok(m));
+        }
+
+        loop {
+            match self.stream.as_mut().poll_next(cx) {
+                Poll::Pending => {
+                    return Poll::Pending;
+                }
+                Poll::Ready(Some(b)) => {
+                    if let Some(m) = self.spare.extend(b.as_ref()) {
+                        self.wake_next();
+                        return Poll::Ready(Ok(m));
+                    }
+                }
+                Poll::Ready(None) => {
+                    return Poll::Ready(Err(Error::EndOfStream {
+                        record_id: RecordId::from(self.next),
+                    }));
+                }
+            }
+        }
+    }
+}
+
+/// Take an ordered stream of bytes and make messages from that stream
+/// available in any order.
+pub struct UnorderedReceiver<S, C>
+where
+    S: Stream<Item = C>,
+    C: AsRef<[u8]>,
+{
+    inner: Arc<Mutex<OperatingState<S, C>>>,
+}
+
+#[allow(dead_code)]
+impl<S, C> UnorderedReceiver<S, C>
+where
+    S: Stream<Item = C>,
+    C: AsRef<[u8]>,
+{
+    /// Wrap a stream for unordered reading.
+    ///
+    /// The capacity here determines how far ahead a read can be.  In most cases,
+    /// this should be the same as the value given to [`ordering_mpsc`].
+    ///
+    /// [`ordering_mpsc`]: crate::helpers::buffers::ordering_mpsc::ordering_mpsc
+    pub fn new(stream: Pin<Box<S>>, capacity: NonZeroUsize) -> Self {
```

Comment:
I don't think that makes this easier in this case.  It would make it less flexible for one (this allows you to read different-sized messages).  This basically takes that functionality and implements it internally, which should be more efficient overall.